openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py

```diff
@@ -0,0 +1,299 @@
+        values that have explicitly been set.
+        """
+        self._check_field(block, name)
+        usage_key = block.scope_ids.usage_id
+        return self.field_data[usage_key][name]
+
+    def set(self, block, name, value):
+        """
+        Set a field for a block to a value.
+        """
+        self._check_field(block, name)
+        usage_key = block.scope_ids.usage_id
+
+        # Check to see if we're just setting the same value. If so, return
+        # without doing anything.
+        block_fields = self.field_data[usage_key]
+        if (name in block_fields) and (block_fields[name] == value):
+            return
+
+        block_fields[name] = value
+        self.changed.add(usage_key)
+
+    def has_changes(self, block):
+        """
+        Does this block have changes that need to be persisted?
+        """
+        return block.scope_ids.usage_id in self.changed
+
+    def _getfield(self, block, name):
+        """
+        Return the field with the given `name` from `block`.
+        If the XBlock doesn't have such a field, raises a KeyError.
+        """
+        # First, get the field from the class, if defined
+        block_field = getattr(block.__class__, name, None)
+        if block_field is not None and isinstance(block_field, Field):
+            return block_field
+        # Not in the class, so name really doesn't name a field
+        raise KeyError(name)
+
+    def _check_field(self, block, name):
+        """
+        Given a block and the name of one of its fields, check that we will be
+        able to read/write it.
+        """
+        field = self._getfield(block, name)
+        if field.scope not in (Scope.content, Scope.settings):
+            raise NotImplementedError(
+                f"Scope {field.scope} (field {name} of {block.scope_ids.usage_id}) "
+                "is unsupported. LearningCoreFieldData only supports the content"
+                " and settings scopes."
+            )
+
+
+class LearningCoreXBlockRuntime(XBlockRuntime):
+    """
+    XBlock runtime that uses openedx-learning apps for content storage.
+
+    The superclass is doing all the hard stuff. This class only only has to
+    worry about the block storage, block serialization/de-serialization, and
+    (eventually) asset storage.
+    """
+
+    def get_block(self, usage_key, for_parent=None):
+        """
+        Fetch an XBlock from Learning Core data models.
+
+        This method will find the OLX for the content in Learning Core, parse it
+        into an XBlock (with mixins) instance, and properly initialize our
+        internal LearningCoreFieldData instance with the field values from the
+        parsed OLX.
+        """
+        # We can do this more efficiently in a single query later, but for now
+        # just get it the easy way.
+        component = self._get_component_from_usage_key(usage_key)
+        component_version = component.versioning.draft
```

Comment:
This runtime needs to support both draft and published versions, and shouldn't be forcing draft version only. (Which one is used should be a property of the runtime when it gets initialized, depending on whether it's running in Studio, LMS, or LMS-Preview.)

We can defer that to a following PR if you want, but if so it needs to be clearly flagged with a TODO.

Reply:
I had figured it would be puntable for a little while because libraries is Studio only, and we're copying to Modulestore before serving to the LMS.

Reply:
Added a TODO comment

Reply:
We did have two use cases for this already, but I think they're both gone now:
1. LabXchange used the "allow_public_learning" feature, which allows users to view library content via the LMS directly, if they know the URL of an individual component. Basically like a "chromeless" view for use in an iframe, but served by the LMS from the published version, and the LMS can save the learner state.
2. The Library Authoring frontend had an "Preview" tab alongside the "OLX" tab and other tabs, which showed the published LMS version in this same way.

1 is no longer used and 2 is removed I think, so it's fine to punt this for now.